### PR TITLE
multiclass val accuracy during finetuning

### PR DESCRIPTION
**Pull request by ramakanth-pasunuru**
Changes:
- remove sampling for validation splits
- update validation accuracy to support multi-label (previously we support only two-label tasks) during finetuning

### CODE REVIEW

File: metaseq/data/streaming_src_tgt_dataset.py
Comment (todpole3) [inline metaseq/data/streaming_src_tgt_dataset.py:237]: Switch to left-truncation now?

File: metaseq/data/streaming_src_tgt_dataset.py
Comment (todpole3) [inline metaseq/data/streaming_src_tgt_dataset.py:261]: This function is a duplicate of yield_src_tgt_single_sentences_pad_8 with two additional fields in each returned element, right? Perhaps be more explicit about this in the description in line 224. Alternatively, can this be implemented in yield_src_tgt_single_sentences_pad_8 with a boolean argument to control whether to add the two additional fields?

File: metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py
Comment (todpole3) [inline metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py:47]: This seems to be an important change, what's the motivation?

File: metaseq/tasks/streaming_finetune_language_modeling.py
Comment (todpole3) [inline metaseq/tasks/streaming_finetune_language_modeling.py:128]: What are the different modes for valid and test?

**Comment by ramakanth-pasunuru** [comment]
Refactored code such that instead of using different break mode for calculating validation accuracy, data reading way is changed.

File: metaseq/data/streaming_src_tgt_dataset.py
Comment (ramakanth-pasunuru) [reply metaseq/data/streaming_src_tgt_dataset.py]: After your comment, I removed relying on different break mode, instead directly change the way we read data for calculating validation accuracy. See my new commits.

File: metaseq/data/streaming_src_tgt_dataset.py
Comment (ramakanth-pasunuru) [reply metaseq/data/streaming_src_tgt_dataset.py]: When they have negative candidates

File: metaseq/data/streaming_src_tgt_dataset.py
Comment (ramakanth-pasunuru) [reply metaseq/data/streaming_src_tgt_dataset.py]: removed this module

File: metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py
Comment (ramakanth-pasunuru) [reply metaseq/model_parallel/criterions/vocab_parallel_cross_entropy.py:47]: mean was added just in my previous PR. We need to see what makes most sense.

File: metaseq/tasks/streaming_finetune_language_modeling.py
Comment (ramakanth-pasunuru) [reply metaseq/tasks/streaming_finetune_language_modeling.py:155]: outdated. No using this anymore

File: metaseq/data/streaming_src_tgt_dataset.py
Comment (ramakanth-pasunuru) [reply metaseq/data/streaming_src_tgt_dataset.py:197]: this module is removed, but it's a good point. Let's decide on this truncation method and update it in a separate PR.

File: metaseq/data/streaming_src_tgt_dataset.py
Comment (stephenroller) [inline metaseq/data/streaming_src_tgt_dataset.py:128]: I'm not sure what's going on here. Can we add some comments explaining why we might get tuples?


Nit: move the comment inside the if block

File: metaseq/data/streaming_src_tgt_dataset.py
Comment (stephenroller) [inline metaseq/data/streaming_src_tgt_dataset.py:186]: In general I feel like we're starting to have a lot of ambiguous types and wondering if we shouldn't subclass for your case, rather than trying to bring more logic into SLM?

**Review by stephenroller — COMMENTED** [review]
Broadly this feels like a lot of project-oriented logic bubbling up into top level abstractions. Let's think about how we could refactor, perhaps by subclassing losses/tasks and providing additional task-specific hooks to call from train.py

**Review by sriniiyer — APPROVED** [review]
LGTM

File: metaseq/data/streaming_src_tgt_dataset.py
Comment (ramakanth-pasunuru) [reply metaseq/data/streaming_src_tgt_dataset.py]: updated.

File: metaseq/data/streaming_src_tgt_dataset.py
Comment (ramakanth-pasunuru) [reply metaseq/data/streaming_src_tgt_dataset.py:186]: Thanks for the suggestions. I think adding negative examples would be in general useful (if not now, later) incase if we implement different losses that consider negative examples as well. So, I feel like SFLM shud have this capability of processing negatives. Probably, the current implementation is not the best to add this feature, but we will improve it later before merging into main branch. Currently, moving on to unblock some experiments.

**Comment by stephenroller** [comment]
Wait no. I didn't want this merged.

**Comment by stephenroller** [comment]
Oh this wasn't merging into Main. Okay no worries.

**Pull request merged**
